File: golang-devildyw/simple/L20有效的括号.go
package main

func main() {

}

// 栈不就是先进先出吗 虽然go没有现成的stack 但用slice也能实现
func isValid(s string) bool {
	// 栈
	stack := make([]byte, 0)
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			// append 函数本来就是往slice后添加元素的
			stack = append(stack, ')')
		} else if s[i] == '{' {
			stack = append(stack, '}')
		} else if s[i] == '[' {
			stack = append(stack, ']')
		} else {

			if len(stack) <= 0 || stack[len(stack)-1] != s[i] {
				return false
			}
			// 出栈
			stack = stack[:len(stack)-1]
		}

	}
	if len(stack) > 0 {
		return false
	}
	return true
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L20有效的括号.java
package top.dyw.leetcode.Hot100;

import java.util.*;

public class L20有效的括号 {
    public boolean isValid(String s) {
        Deque<Character> stack = new LinkedList<>();

        for(char c : s.toCharArray()) {
            if (c=='(') {
                stack.push(')');
            } else if (c=='{') {
                stack.push('}');
            } else if (c=='[') {
                stack.push(']');
            } else {
                if (stack.isEmpty()) {
                    return false;
                } else {
                    if (stack.peek()==c) {
                        stack.pop();
                    } else {
                        break;
                    }
                }
            }
        }

        if (!stack.isEmpty()){
            return false;
        }
        return true;
    }
}
